src/opencl/pgpsda_kernel.cl

```diff
@@ -0,0 +1,74 @@
+/*
+ * This software is Copyright (c) 2017 Dhiru Kholia <dhiru at openwall.com> and
+ * Copyright (c) 2017 magnum, and
+ * it is hereby released to the general public under the following terms:
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted.
+ */
+
+#include "opencl_device_info.h"
+#include "opencl_misc.h"
+#include "opencl_sha1_ctx.h"
+
+#ifndef PLAINTEXT_LENGTH
+#error PLAINTEXT_LENGTH must be defined
+#endif
+
+#ifndef SHA1_DIGEST_LENGTH
+#define SHA1_DIGEST_LENGTH 20
+#endif
+
+typedef struct {
+	uint length;
+	uchar v[PLAINTEXT_LENGTH];
+} pgpsda_password;
+
+typedef struct {
+	uchar v[16];
+} pgpsda_hash;
+
+typedef struct {
+	uint iterations;
+	uchar salt[8];
+} pgpsda_salt;
+
+#ifndef __MESA__
+inline
+#endif
+void pgpsda_kdf(__global const uchar *ipassword, const uint plen,
+                 __constant uchar *isalt, const uint iterations,
+				 __global uchar *okey)
+{
+	SHA_CTX ctx;
+	uint32_t i = 0, j = 0; // These variables are of type uint16_t in the original code
+	unsigned char password[PLAINTEXT_LENGTH];
+	unsigned char salt[8];
+	unsigned char key[SHA1_DIGEST_LENGTH];
+	memcpy_cp(salt, isalt, 8);
+	memcpy_gp(password, ipassword, plen);
+
+	SHA1_Init(&ctx);
+	SHA1_Update(&ctx, salt, 8);
+	for (i = 0, j = 0; i < iterations; i++, j++) {
+		SHA1_Update(&ctx, password, plen);
+#if __ENDIAN_LITTLE__
+		SHA1_Update(&ctx, (uchar*)&j, 1);
+#else
+		SHA1_Update(&ctx, ((uchar*)&j) + 3, 1);
+#endif
+	}
```

Comment:
`i` and `j` are the same at all times. Just drop `j` and use `i`. The same goes for CPU code (and pgpdisk_fmt_plug.c also)

Reply:
I noticed this earlier today.

Here is the original "upstream" code.

```
PGPUInt8    j;
PGPUInt16  i, k;

for (i=0, j=0; i<k; i++, j++)
{
    pgpSHAUpdate(&sha,gpp->pszPassPhrase,strlen(gpp->pszPassPhrase));   
    pgpSHAUpdate(&sha,&j,1);
}
```

I was wrong when I said that `i` and `j` are unsigned 16-bit numbers. After double-checking the original code, `j` is an unsigned char. It seems that integer wrapping would come info affect with the original code. I wondering how our code (which now uses 32-bit numbers) is working well too.

Reply:
Hmm. I think that in our code (which now uses 32-bit numbers), the lowest byte is affected in the same way as the original `unsigned char j` is affected by the increment operations. I am still a bit uneasy about all this :-)

Reply:
This original code has a comment saying `Hash the passphrase and a rotating byte counter the specified number of times`.

Reply:
To convince myself that things are OK in our code, I wrote the following code,

``` C
#include <stdio.h>
#include <stdint.h>

int main()
{
	uint8_t j; // original j
	uint16_t iterations = 16000;
	uint16_t i;

	uint32_t oj; // our 32-bit j

	for (i = 0, j = 0, oj = 0; i < iterations; i++, j++, oj++) {
		uint8_t *v = (uint8_t *)&i;

		if (j != *v) {
			printf("Failed!\n");
			break;
		}
		printf("%d ", j);
	}
	printf("\n");

	return 0;
}
```